driver/src/price_finding/optimization_price_finder.rs

```diff
@@ -119,116 +206,34 @@ impl OptimisationPriceFinder {
     }
 }
 
-fn token_id(token: u16) -> String {
-    format!("T{:04}", token)
-}
-
-fn account_id(account: H160) -> String {
-    format!("{:x}", account)
-}
-
-fn serialize_tokens(orders: &[models::Order]) -> Vec<String> {
+fn serialize_tokens(orders: &[models::Order]) -> BTreeSet<TokenId> {
```

Comment:
Why do we use the BtreeSet here? Wouldn't a hashmap perform better?

Reply:
The solver expects a token collection that is in order and without duplicate token identifiers. Using a `BTreeSet` does both these things for in a way that is guaranteed by the data structure. Additionally, the JSON serialized array that this produces will also always be in token ID order and does not require any extra serialization code, or ensuring that the `Vec` is sorted and only contains unique elements before serializing.

The main motivation was purely to have a data structure that reflects exactly how the data is meant to look like. In terms of performance, I don't think there is a significant difference. Note that sorting a `Vec` is `O(n*log(n))` and inserting `n` items into a `BTreeSet` is also `O(n*log(n))` so, asymptotically, they should perform the same.

Reply:
Also, places like here: https://github.com/gnosis/dex-services/blob/3321a911e655c4a47be657bc1c5611823ff17d20/driver/src/price_finding/optimization_price_finder.rs#L169

I changed from a `HashMap` to a `BTreeMap` as the data is expected to be serialized in account ID order. While a `HashMap` would perform better for insertion and random access, when serializing we were collecting the `HashMap` into a `BTreeMap` to ensure that it would get serialized in order. Since the `solver_input::Input` type is only used for serialization, I thought it would be better to just create it with the `BTreeMap` directly and skip creating a `HashMap` only to create a `BTreeMap` afterwards.

If we were using this data internally in the driver as well, then using a `HashMap` would probably have performed better as you suggested.